File: C++/tree_time_traversal.cpp
#include <chrono>
#include <iostream>
#include <unistd.h>
#include <queue>

using namespace std;

class Node {
public:
    Node(Node* p) {
        parent = p;
        this->data = 0;
        this->leftChild = nullptr;
        this->rightChild = nullptr;
    }

    Node* parent;
    Node* leftChild;
    Node* rightChild;
    int data;

    void DFS() {
        data += 1;
        if (leftChild != nullptr) { leftChild->DFS(); }
        if (rightChild != nullptr) {rightChild->DFS(); }
    }
};

class Tree {
public:
    Tree(int nodeNum) {
        rootNode = new Node(nullptr);
        queue<Node*> nodeQueue;
        nodeQueue.push(rootNode);
        for (int i = 0; i < nodeNum; i++) {
            Node* currentNode = nodeQueue.front();
            nodeQueue.pop();
            currentNode->leftChild = new Node(currentNode);
            currentNode->rightChild = new Node(currentNode);
            nodeQueue.push(currentNode->leftChild);
            nodeQueue.push(currentNode->rightChild);

        }
    }

    void DFS() {
        rootNode->DFS();
    }
private:
    Node* rootNode;
};

int main() {
    const int NUM_TRAILS = 50;
    const int ONE_BILLION = 1000000000;
    const int ONE_MILLION = 1000000;
    std::chrono::high_resolution_clock m_clock;
    double cumulative = 0;
    Tree* tree = new Tree(ONE_MILLION);
    for (int t = 0; t < NUM_TRAILS; t++) {
        uint64_t startTime = std::chrono::duration_cast<std::chrono::nanoseconds>(m_clock.now().time_since_epoch()).count();
        tree->DFS();
        uint64_t endTime = std::chrono::duration_cast<std::chrono::nanoseconds>(m_clock.now().time_since_epoch()).count();

        uint64_t difference = endTime - startTime;
        cumulative = cumulative + (double)difference;
        double seconds = difference / (double) ONE_BILLION;

        cout << "Time to process: " << seconds << " seconds" << '\n';
        sleep(0.003);
    }
    double seconds = cumulative / ((double) NUM_TRAILS) / ((double) ONE_BILLION);
    cout << "Total time to process: " << seconds << " seconds" << '\n';
}

File: Python/tree_time_traversal.py
import time

ONE_MILLION = 1000000
ONE_BILLION = 1000000000
NUM_TRIALS = 50
cumulative = 0

class Tree:
    def __init__(self, nodeNum):
        self._root = Node(None)
        node_queue = []
        node_queue.append(self._root)
        for i in range(nodeNum):
            current_node = node_queue.pop(0)
            current_node._left_child = Node(current_node)
            current_node._right_child = Node(current_node)
            node_queue.append(current_node._left_child)
            node_queue.append(current_node._right_child)

    def DFS(self):
        self._root.DFS()


class Node:
    def __init__(self, p):
        self._left_child = None
        self._right_child = None
        self._parent = p
        self._data = 0

    def DFS(self):
        self._data += 1
        if self._left_child != None:
            self._left_child.DFS()
        if self._right_child != None:
            self._right_child.DFS()
    


tree = Tree(ONE_MILLION)

NUM_TRIALS = 50
cumulative = 0

for t in range(NUM_TRIALS):
    startTime = time.time_ns()
    tree.DFS()
    endTime = time.time_ns()
    difference = endTime - startTime
    cumulative += difference
    seconds = difference / ONE_BILLION
    print(f"Time to process: {seconds} seconds")
    time.sleep(0.0003)

seconds = cumulative / ONE_BILLION / NUM_TRIALS
print(f"Average time to process: {seconds} seconds")